crates/pathfinder/src/state/contract_hash.rs

```diff
@@ -0,0 +1,414 @@
+        /// Keccak256 hash.
+        #[serde(skip_serializing, borrow)]
+        pub entry_points_by_type: HashMap<EntryPointType, Vec<SelectorAndOffset<'a>>>,
+    }
+
+    #[derive(Copy, Clone, Debug, serde::Deserialize, PartialEq, Hash, Eq)]
+    #[serde(deny_unknown_fields)]
+    pub enum EntryPointType {
+        #[serde(rename = "EXTERNAL")]
+        External,
+        #[serde(rename = "L1_HANDLER")]
+        L1Handler,
+        #[serde(rename = "CONSTRUCTOR")]
+        Constructor,
+    }
+
+    impl fmt::Display for EntryPointType {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            use EntryPointType::*;
+            f.pad(match self {
+                External => "EXTERNAL",
+                L1Handler => "L1_HANDLER",
+                Constructor => "CONSTRUCTOR",
+            })
+        }
+    }
+
+    #[derive(serde::Deserialize)]
+    #[serde(deny_unknown_fields)]
+    pub struct SelectorAndOffset<'a> {
+        pub selector: Cow<'a, str>,
+        pub offset: Cow<'a, str>,
+    }
+
+    // It's important that this is ordered alphabetically because the fields need to be in
+    // sorted order for the keccak hashed representation.
+    #[derive(serde::Deserialize, serde::Serialize)]
+    #[serde(deny_unknown_fields)]
+    pub struct Program<'a> {
+        #[serde(skip_serializing_if = "Vec::is_empty")]
+        pub attributes: Vec<serde_json::Value>,
+
+        pub builtins: Vec<Cow<'a, str>>,
+
+        pub data: Vec<Cow<'a, str>>,
+
+        #[serde(borrow)]
+        pub debug_info: Option<&'a serde_json::value::RawValue>,
+
+        // Important that this is ordered by the numeric keys, not lexicographically
+        pub hints: BTreeMap<u64, Vec<serde_json::Value>>,
+
+        pub identifiers: serde_json::Value,
+
+        #[serde(borrow)]
+        pub main_scope: Cow<'a, str>,
+
+        // Unlike most other integers, this one is hex string. We don't need to interpret it,
+        // it just needs to be part of the hashed output.
+        #[serde(borrow)]
+        pub prime: Cow<'a, str>,
+
+        pub reference_manager: serde_json::Value,
+    }
+
+    #[cfg(test)]
+    mod roundtrip_tests {
+        // FIXME: we should have many test cases utilizing this.
+        #[allow(unused)]
+        fn roundtrips<'a, T>(input: &'a str)
+        where
+            T: serde::Deserialize<'a> + serde::Serialize,
+        {
+            use super::super::PythonDefaultFormatter;
+
+            let parsed: T = serde_json::from_str(input).unwrap();
+            let mut ser =
+                serde_json::Serializer::with_formatter(Vec::new(), PythonDefaultFormatter);
+            parsed.serialize(&mut ser).unwrap();
+            let bytes = ser.into_inner();
+            let output = std::str::from_utf8(&bytes).expect("serde does this unchecked");
+
+            // these need to be byte for byte equal because we hash this
+            assert_eq!(input, output);
+        }
```

Comment:
Not realistic I could come up with something, however on the first bug, this should be used to start recording those cases.